File: sdks/go/client_test.go
package streamhouse

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

// Helper to create a test server
func newTestServer(t *testing.T, handler http.HandlerFunc) (*httptest.Server, *Client) {
	server := httptest.NewServer(handler)
	client := NewClient(server.URL)
	return server, client
}

func TestNewClient(t *testing.T) {
	client := NewClient("http://localhost:8080")
	if client.baseURL != "http://localhost:8080" {
		t.Errorf("expected baseURL 'http://localhost:8080', got '%s'", client.baseURL)
	}
	if client.apiURL != "http://localhost:8080/api/v1" {
		t.Errorf("expected apiURL 'http://localhost:8080/api/v1', got '%s'", client.apiURL)
	}
}

func TestNewClientWithTrailingSlash(t *testing.T) {
	client := NewClient("http://localhost:8080/")
	if client.baseURL != "http://localhost:8080" {
		t.Errorf("expected baseURL without trailing slash, got '%s'", client.baseURL)
	}
}

func TestNewClientWithOptions(t *testing.T) {
	client := NewClient("http://localhost:8080",
		WithAPIKey("test-key"),
		WithTimeout(60*time.Second),
	)
	if client.apiKey != "test-key" {
		t.Errorf("expected apiKey 'test-key', got '%s'", client.apiKey)
	}
	if client.httpClient.Timeout != 60*time.Second {
		t.Errorf("expected timeout 60s, got '%v'", client.httpClient.Timeout)
	}
}

func TestListTopics(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/topics" {
			t.Errorf("expected path '/api/v1/topics', got '%s'", r.URL.Path)
		}
		if r.Method != http.MethodGet {
			t.Errorf("expected method GET, got '%s'", r.Method)
		}

		topics := []Topic{
			{Name: "topic1", Partitions: 3, ReplicationFactor: 1},
			{Name: "topic2", Partitions: 6, ReplicationFactor: 2},
		}
		json.NewEncoder(w).Encode(topics)
	})
	defer server.Close()

	ctx := context.Background()
	topics, err := client.ListTopics(ctx)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(topics) != 2 {
		t.Errorf("expected 2 topics, got %d", len(topics))
	}
	if topics[0].Name != "topic1" {
		t.Errorf("expected topic name 'topic1', got '%s'", topics[0].Name)
	}
}

func TestCreateTopic(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/topics" {
			t.Errorf("expected path '/api/v1/topics', got '%s'", r.URL.Path)
		}
		if r.Method != http.MethodPost {
			t.Errorf("expected method POST, got '%s'", r.Method)
		}

		var req createTopicRequest
		json.NewDecoder(r.Body).Decode(&req)

		if req.Name != "my-topic" {
			t.Errorf("expected name 'my-topic', got '%s'", req.Name)
		}
		if req.Partitions != 3 {
			t.Errorf("expected partitions 3, got %d", req.Partitions)
		}

		topic := Topic{
			Name:              req.Name,
			Partitions:        req.Partitions,
			ReplicationFactor: req.ReplicationFactor,
			CreatedAt:         "2024-01-01T00:00:00Z",
		}
		w.WriteHeader(http.StatusCreated)
		json.NewEncoder(w).Encode(topic)
	})
	defer server.Close()

	ctx := context.Background()
	topic, err := client.CreateTopic(ctx, "my-topic", 3)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if topic.Name != "my-topic" {
		t.Errorf("expected topic name 'my-topic', got '%s'", topic.Name)
	}
	if topic.Partitions != 3 {
		t.Errorf("expected 3 partitions, got %d", topic.Partitions)
	}
}

func TestCreateTopicWithReplicationFactor(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		var req createTopicRequest
		json.NewDecoder(r.Body).Decode(&req)

		if req.ReplicationFactor != 2 {
			t.Errorf("expected replication factor 2, got %d", req.ReplicationFactor)
		}

		topic := Topic{
			Name:              req.Name,
			Partitions:        req.Partitions,
			ReplicationFactor: req.ReplicationFactor,
		}
		json.NewEncoder(w).Encode(topic)
	})
	defer server.Close()

	ctx := context.Background()
	topic, err := client.CreateTopic(ctx, "my-topic", 3, WithReplicationFactor(2))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if topic.ReplicationFactor != 2 {
		t.Errorf("expected replication factor 2, got %d", topic.ReplicationFactor)
	}
}

func TestGetTopic(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/topics/my-topic" {
			t.Errorf("expected path '/api/v1/topics/my-topic', got '%s'", r.URL.Path)
		}

		topic := Topic{Name: "my-topic", Partitions: 3}
		json.NewEncoder(w).Encode(topic)
	})
	defer server.Close()

	ctx := context.Background()
	topic, err := client.GetTopic(ctx, "my-topic")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if topic.Name != "my-topic" {
		t.Errorf("expected topic name 'my-topic', got '%s'", topic.Name)
	}
}

func TestDeleteTopic(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/topics/my-topic" {
			t.Errorf("expected path '/api/v1/topics/my-topic', got '%s'", r.URL.Path)
		}
		if r.Method != http.MethodDelete {
			t.Errorf("expected method DELETE, got '%s'", r.Method)
		}
		w.WriteHeader(http.StatusNoContent)
	})
	defer server.Close()

	ctx := context.Background()
	err := client.DeleteTopic(ctx, "my-topic")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestProduce(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/produce" {
			t.Errorf("expected path '/api/v1/produce', got '%s'", r.URL.Path)
		}

		var req produceRequest
		json.NewDecoder(r.Body).Decode(&req)

		if req.Topic != "events" {
			t.Errorf("expected topic 'events', got '%s'", req.Topic)
		}
		if req.Value != `{"event":"click"}` {
			t.Errorf("unexpected value: %s", req.Value)
		}

		result := ProduceResult{Offset: 42, Partition: 0}
		json.NewEncoder(w).Encode(result)
	})
	defer server.Close()

	ctx := context.Background()
	result, err := client.Produce(ctx, "events", `{"event":"click"}`)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Offset != 42 {
		t.Errorf("expected offset 42, got %d", result.Offset)
	}
}

func TestProduceWithKey(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		var req produceRequest
		json.NewDecoder(r.Body).Decode(&req)

		if req.Key == nil || *req.Key != "user-123" {
			t.Errorf("expected key 'user-123', got %v", req.Key)
		}

		result := ProduceResult{Offset: 42, Partition: 0}
		json.NewEncoder(w).Encode(result)
	})
	defer server.Close()

	ctx := context.Background()
	_, err := client.Produce(ctx, "events", `{"event":"click"}`, WithKey("user-123"))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestProduceBatch(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/produce/batch" {
			t.Errorf("expected path '/api/v1/produce/batch', got '%s'", r.URL.Path)
		}

		var req batchProduceRequest
		json.NewDecoder(r.Body).Decode(&req)

		if len(req.Records) != 2 {
			t.Errorf("expected 2 records, got %d", len(req.Records))
		}

		result := BatchProduceResult{
			Count: 2,
			Offsets: []BatchRecordResult{
				{Partition: 0, Offset: 10},
				{Partition: 0, Offset: 11},
			},
		}
		json.NewEncoder(w).Encode(result)
	})
	defer server.Close()

	ctx := context.Background()
	result, err := client.ProduceBatch(ctx, "events", []BatchRecord{
		NewBatchRecord(`{"event":"a"}`),
		NewBatchRecordWithKey(`{"event":"b"}`, "key1"),
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Count != 2 {
		t.Errorf("expected count 2, got %d", result.Count)
	}
}

func TestConsume(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/consume" {
			t.Errorf("expected path '/api/v1/consume', got '%s'", r.URL.Path)
		}

		query := r.URL.Query()
		if query.Get("topic") != "events" {
			t.Errorf("expected topic 'events', got '%s'", query.Get("topic"))
		}
		if query.Get("partition") != "0" {
			t.Errorf("expected partition '0', got '%s'", query.Get("partition"))
		}

		result := ConsumeResult{
			Records: []ConsumedRecord{
				{Partition: 0, Offset: 0, Value: `{"event":"a"}`},
				{Partition: 0, Offset: 1, Value: `{"event":"b"}`},
			},
			NextOffset: 2,
		}
		json.NewEncoder(w).Encode(result)
	})
	defer server.Close()

	ctx := context.Background()
	result, err := client.Consume(ctx, "events", 0, ConsumeOptions{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(result.Records) != 2 {
		t.Errorf("expected 2 records, got %d", len(result.Records))
	}
	if result.NextOffset != 2 {
		t.Errorf("expected next offset 2, got %d", result.NextOffset)
	}
}

func TestQuery(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/sql" {
			t.Errorf("expected path '/api/v1/sql', got '%s'", r.URL.Path)
		}

		var req sqlQueryRequest
		json.NewDecoder(r.Body).Decode(&req)

		if req.Query != "SELECT * FROM events" {
			t.Errorf("unexpected query: %s", req.Query)
		}

		result := SQLResult{
			Columns: []ColumnInfo{
				{Name: "offset", DataType: "bigint"},
				{Name: "value", DataType: "string"},
			},
			Rows:            []json.RawMessage{[]byte(`[0, "test"]`)},
			RowCount:        1,
			ExecutionTimeMs: 50,
			Truncated:       false,
		}
		json.NewEncoder(w).Encode(result)
	})
	defer server.Close()

	ctx := context.Background()
	result, err := client.Query(ctx, "SELECT * FROM events", QueryOptions{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.RowCount != 1 {
		t.Errorf("expected row count 1, got %d", result.RowCount)
	}
	if len(result.Columns) != 2 {
		t.Errorf("expected 2 columns, got %d", len(result.Columns))
	}
}

func TestHealthCheck(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/health" {
			w.WriteHeader(http.StatusNotFound)
			return
		}
		w.WriteHeader(http.StatusOK)
		w.Write([]byte(`{"status":"ok"}`))
	})
	defer server.Close()

	ctx := context.Background()
	healthy := client.HealthCheck(ctx)
	if !healthy {
		t.Error("expected health check to pass")
	}
}

func TestHealthCheckFailed(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusServiceUnavailable)
	})
	defer server.Close()

	ctx := context.Background()
	healthy := client.HealthCheck(ctx)
	if healthy {
		t.Error("expected health check to fail")
	}
}

// Error handling tests
func TestNotFoundError(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		w.Write([]byte(`{"error":"topic not found"}`))
	})
	defer server.Close()

	ctx := context.Background()
	_, err := client.GetTopic(ctx, "nonexistent")
	if err == nil {
		t.Fatal("expected error")
	}
	if !IsNotFound(err) {
		t.Errorf("expected NotFoundError, got %T", err)
	}
}

func TestValidationError(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte(`{"error":"invalid partitions"}`))
	})
	defer server.Close()

	ctx := context.Background()
	_, err := client.CreateTopic(ctx, "test", 0)
	if err == nil {
		t.Fatal("expected error")
	}
	if !IsValidation(err) {
		t.Errorf("expected ValidationError, got %T", err)
	}
}

func TestConflictError(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusConflict)
		w.Write([]byte(`{"error":"topic already exists"}`))
	})
	defer server.Close()

	ctx := context.Background()
	_, err := client.CreateTopic(ctx, "existing", 3)
	if err == nil {
		t.Fatal("expected error")
	}
	if !IsConflict(err) {
		t.Errorf("expected ConflictError, got %T", err)
	}
}

func TestServerError(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusInternalServerError)
		w.Write([]byte(`{"error":"internal error"}`))
	})
	defer server.Close()

	ctx := context.Background()
	_, err := client.ListTopics(ctx)
	if err == nil {
		t.Fatal("expected error")
	}
	if !IsServer(err) {
		t.Errorf("expected ServerError, got %T", err)
	}
}

func TestAPIKeyHeader(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		auth := r.Header.Get("Authorization")
		if auth != "Bearer test-api-key" {
			t.Errorf("expected Authorization header 'Bearer test-api-key', got '%s'", auth)
		}
		json.NewEncoder(w).Encode([]Topic{})
	})
	defer server.Close()

	client.apiKey = "test-api-key"
	ctx := context.Background()
	_, err := client.ListTopics(ctx)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestContextCancellation(t *testing.T) {
	server, client := newTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(100 * time.Millisecond)
		json.NewEncoder(w).Encode([]Topic{})
	})
	defer server.Close()

	ctx, cancel := context.WithCancel(context.Background())
	cancel() // Cancel immediately

	_, err := client.ListTopics(ctx)
	if err == nil {
		t.Fatal("expected error due to cancelled context")
	}
}

File: sdks/typescript/src/client.test.ts
/**
 * Tests for StreamHouse TypeScript client.
 */

import { describe, it, expect, vi, beforeEach } from 'vitest';
import { StreamHouseClient } from './client';
import {
  NotFoundError,
  ValidationError,
  ConflictError,
  ServerError,
  AuthenticationError,
  TimeoutError,
  ConnectionError,
} from './errors';

// Mock fetch response helper
function mockResponse(status: number, body?: unknown): Response {
  return {
    ok: status >= 200 && status < 300,
    status,
    json: () => Promise.resolve(body),
    text: () => Promise.resolve(body ? JSON.stringify(body) : ''),
  } as Response;
}

describe('StreamHouseClient', () => {
  let mockFetch: ReturnType<typeof vi.fn>;

  beforeEach(() => {
    mockFetch = vi.fn();
  });

  describe('Client Initialization', () => {
    it('should initialize with string URL', () => {
      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      expect(client).toBeDefined();
    });

    it('should strip trailing slash from URL', () => {
      mockFetch.mockResolvedValue(mockResponse(200, []));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080/',
        fetch: mockFetch,
      });

      client.listTopics();

      expect(mockFetch).toHaveBeenCalledWith(
        'http://localhost:8080/api/v1/topics',
        expect.any(Object)
      );
    });

    it('should accept string as constructor argument', () => {
      // This will throw because no fetch is provided in global context during tests
      expect(() => new StreamHouseClient('http://localhost:8080')).toThrow();
    });
  });

  describe('Topic Operations', () => {
    it('should list topics', async () => {
      const topics = [
        { name: 'topic1', partitions: 3, replicationFactor: 1, createdAt: '2024-01-01' },
        { name: 'topic2', partitions: 6, replicationFactor: 2, createdAt: '2024-01-02' },
      ];
      mockFetch.mockResolvedValue(mockResponse(200, topics));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const result = await client.listTopics();

      expect(result).toHaveLength(2);
      expect(result[0].name).toBe('topic1');
      expect(result[1].partitions).toBe(6);
    });

    it('should create a topic', async () => {
      const topic = {
        name: 'my-topic',
        partitions: 3,
        replicationFactor: 1,
        createdAt: '2024-01-01',
      };
      mockFetch.mockResolvedValue(mockResponse(201, topic));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const result = await client.createTopic('my-topic', 3);

      expect(result.name).toBe('my-topic');
      expect(result.partitions).toBe(3);

      expect(mockFetch).toHaveBeenCalledWith(
        'http://localhost:8080/api/v1/topics',
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({
            name: 'my-topic',
            partitions: 3,
            replication_factor: 1,
          }),
        })
      );
    });

    it('should get a topic', async () => {
      const topic = { name: 'my-topic', partitions: 3 };
      mockFetch.mockResolvedValue(mockResponse(200, topic));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const result = await client.getTopic('my-topic');

      expect(result.name).toBe('my-topic');
      expect(mockFetch).toHaveBeenCalledWith(
        'http://localhost:8080/api/v1/topics/my-topic',
        expect.any(Object)
      );
    });

    it('should delete a topic', async () => {
      mockFetch.mockResolvedValue(mockResponse(204));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      await client.deleteTopic('my-topic');

      expect(mockFetch).toHaveBeenCalledWith(
        'http://localhost:8080/api/v1/topics/my-topic',
        expect.objectContaining({ method: 'DELETE' })
      );
    });

    it('should list partitions', async () => {
      const partitions = [
        { topic: 'events', partitionId: 0, highWatermark: 100 },
        { topic: 'events', partitionId: 1, highWatermark: 200 },
      ];
      mockFetch.mockResolvedValue(mockResponse(200, partitions));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const result = await client.listPartitions('events');

      expect(result).toHaveLength(2);
    });

    it('should encode topic names in URLs', async () => {
      mockFetch.mockResolvedValue(mockResponse(200, { name: 'my topic' }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      await client.getTopic('my topic');

      expect(mockFetch).toHaveBeenCalledWith(
        'http://localhost:8080/api/v1/topics/my%20topic',
        expect.any(Object)
      );
    });
  });

  describe('Producer Operations', () => {
    it('should produce a message', async () => {
      const result = { offset: 42, partition: 0 };
      mockFetch.mockResolvedValue(mockResponse(200, result));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const produceResult = await client.produce('events', '{"event": "click"}');

      expect(produceResult.offset).toBe(42);
      expect(produceResult.partition).toBe(0);
    });

    it('should produce a message with key', async () => {
      const result = { offset: 42, partition: 0 };
      mockFetch.mockResolvedValue(mockResponse(200, result));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      await client.produce('events', '{"event": "click"}', { key: 'user-123' });

      expect(mockFetch).toHaveBeenCalledWith(
        expect.any(String),
        expect.objectContaining({
          body: JSON.stringify({
            topic: 'events',
            value: '{"event": "click"}',
            key: 'user-123',
            partition: undefined,
          }),
        })
      );
    });

    it('should produce a batch of messages', async () => {
      const result = {
        count: 2,
        offsets: [
          { partition: 0, offset: 10 },
          { partition: 0, offset: 11 },
        ],
      };
      mockFetch.mockResolvedValue(mockResponse(200, result));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const batchResult = await client.produceBatch('events', [
        '{"event": "a"}',
        { key: 'user-1', value: '{"event": "b"}' },
      ]);

      expect(batchResult.count).toBe(2);
      expect(batchResult.offsets).toHaveLength(2);
    });
  });

  describe('Consumer Operations', () => {
    it('should consume messages', async () => {
      const result = {
        records: [
          { partition: 0, offset: 0, key: null, value: '{"event": "a"}', timestamp: 1000 },
          { partition: 0, offset: 1, key: 'k1', value: '{"event": "b"}', timestamp: 1001 },
        ],
        nextOffset: 2,
      };
      mockFetch.mockResolvedValue(mockResponse(200, result));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const consumeResult = await client.consume('events', 0);

      expect(consumeResult.records).toHaveLength(2);
      expect(consumeResult.nextOffset).toBe(2);
      expect(consumeResult.records[0].value).toBe('{"event": "a"}');
    });

    it('should consume with options', async () => {
      mockFetch.mockResolvedValue(mockResponse(200, { records: [], nextOffset: 0 }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      await client.consume('events', 0, { offset: 10, maxRecords: 100 });

      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining('offset=10'),
        expect.any(Object)
      );
      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining('maxRecords=100'),
        expect.any(Object)
      );
    });
  });

  describe('Consumer Group Operations', () => {
    it('should list consumer groups', async () => {
      const groups = [
        { groupId: 'group1', topics: ['events'], totalLag: 0 },
        { groupId: 'group2', topics: ['logs'], totalLag: 10 },
      ];
      mockFetch.mockResolvedValue(mockResponse(200, groups));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const result = await client.listConsumerGroups();

      expect(result).toHaveLength(2);
    });

    it('should get consumer group details', async () => {
      const group = {
        groupId: 'my-group',
        offsets: [
          { topic: 'events', partitionId: 0, committedOffset: 50 },
        ],
      };
      mockFetch.mockResolvedValue(mockResponse(200, group));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const result = await client.getConsumerGroup('my-group');

      expect(result.groupId).toBe('my-group');
      expect(result.offsets).toHaveLength(1);
    });

    it('should commit offset', async () => {
      mockFetch.mockResolvedValue(mockResponse(200, { success: true }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const result = await client.commitOffset('my-group', 'events', 0, 42);

      expect(result).toBe(true);
    });

    it('should delete consumer group', async () => {
      mockFetch.mockResolvedValue(mockResponse(204));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      await client.deleteConsumerGroup('my-group');

      expect(mockFetch).toHaveBeenCalledWith(
        'http://localhost:8080/api/v1/consumer-groups/my-group',
        expect.objectContaining({ method: 'DELETE' })
      );
    });
  });

  describe('SQL Operations', () => {
    it('should execute a SQL query', async () => {
      const result = {
        columns: [
          { name: 'offset', dataType: 'bigint' },
          { name: 'value', dataType: 'string' },
        ],
        rows: [[0, 'test']],
        rowCount: 1,
        executionTimeMs: 50,
        truncated: false,
      };
      mockFetch.mockResolvedValue(mockResponse(200, result));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const queryResult = await client.query('SELECT * FROM events');

      expect(queryResult.rowCount).toBe(1);
      expect(queryResult.columns).toHaveLength(2);
      expect(queryResult.columns[0].name).toBe('offset');
    });

    it('should pass timeout option to SQL query', async () => {
      mockFetch.mockResolvedValue(mockResponse(200, { columns: [], rows: [], rowCount: 0 }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      await client.query('SELECT * FROM events', { timeoutMs: 5000 });

      expect(mockFetch).toHaveBeenCalledWith(
        expect.any(String),
        expect.objectContaining({
          body: JSON.stringify({
            query: 'SELECT * FROM events',
            timeout_ms: 5000,
          }),
        })
      );
    });
  });

  describe('Error Handling', () => {
    it('should throw NotFoundError for 404', async () => {
      mockFetch.mockResolvedValue(mockResponse(404, { error: 'topic not found' }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });

      await expect(client.getTopic('nonexistent')).rejects.toThrow(NotFoundError);
    });

    it('should throw ValidationError for 400', async () => {
      mockFetch.mockResolvedValue(mockResponse(400, { error: 'invalid partitions' }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });

      await expect(client.createTopic('test', 0)).rejects.toThrow(ValidationError);
    });

    it('should throw ConflictError for 409', async () => {
      mockFetch.mockResolvedValue(mockResponse(409, { error: 'topic already exists' }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });

      await expect(client.createTopic('existing', 3)).rejects.toThrow(ConflictError);
    });

    it('should throw AuthenticationError for 401', async () => {
      mockFetch.mockResolvedValue(mockResponse(401, { error: 'unauthorized' }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });

      await expect(client.listTopics()).rejects.toThrow(AuthenticationError);
    });

    it('should throw ServerError for 5xx', async () => {
      mockFetch.mockResolvedValue(mockResponse(500, { error: 'internal server error' }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });

      await expect(client.listTopics()).rejects.toThrow(ServerError);
    });

    it('should include error message from server', async () => {
      mockFetch.mockResolvedValue(mockResponse(404, { error: 'topic "test" not found' }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });

      try {
        await client.getTopic('test');
        expect.fail('Should have thrown');
      } catch (error) {
        expect(error).toBeInstanceOf(NotFoundError);
        expect((error as NotFoundError).message).toContain('topic "test" not found');
        expect((error as NotFoundError).statusCode).toBe(404);
      }
    });

    it('should throw ConnectionError on network failure', async () => {
      mockFetch.mockRejectedValue(new Error('Network error'));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });

      await expect(client.listTopics()).rejects.toThrow(ConnectionError);
    });

    it('should throw TimeoutError on abort', async () => {
      mockFetch.mockRejectedValue(Object.assign(new Error('Aborted'), { name: 'AbortError' }));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
        timeout: 1,
      });

      await expect(client.listTopics()).rejects.toThrow(TimeoutError);
    });
  });

  describe('Authentication', () => {
    it('should include API key in Authorization header', async () => {
      mockFetch.mockResolvedValue(mockResponse(200, []));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        apiKey: 'test-api-key',
        fetch: mockFetch,
      });
      await client.listTopics();

      expect(mockFetch).toHaveBeenCalledWith(
        expect.any(String),
        expect.objectContaining({
          headers: expect.objectContaining({
            Authorization: 'Bearer test-api-key',
          }),
        })
      );
    });

    it('should not include Authorization header when no API key', async () => {
      mockFetch.mockResolvedValue(mockResponse(200, []));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      await client.listTopics();

      const callArgs = mockFetch.mock.calls[0][1];
      expect(callArgs.headers.Authorization).toBeUndefined();
    });
  });

  describe('Health Check', () => {
    it('should return true when server is healthy', async () => {
      mockFetch.mockResolvedValue(mockResponse(200));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const healthy = await client.healthCheck();

      expect(healthy).toBe(true);
      expect(mockFetch).toHaveBeenCalledWith(
        'http://localhost:8080/health',
        expect.objectContaining({ method: 'GET' })
      );
    });

    it('should return false when server is unhealthy', async () => {
      mockFetch.mockResolvedValue(mockResponse(503));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const healthy = await client.healthCheck();

      expect(healthy).toBe(false);
    });

    it('should return false on connection error', async () => {
      mockFetch.mockRejectedValue(new Error('Connection refused'));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const healthy = await client.healthCheck();

      expect(healthy).toBe(false);
    });
  });

  describe('Agent Operations', () => {
    it('should list agents', async () => {
      const agents = [
        { agentId: 'agent-1', address: '10.0.0.1:8080' },
        { agentId: 'agent-2', address: '10.0.0.2:8080' },
      ];
      mockFetch.mockResolvedValue(mockResponse(200, agents));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const result = await client.listAgents();

      expect(result).toHaveLength(2);
    });

    it('should get metrics', async () => {
      const metrics = {
        topicsCount: 5,
        agentsCount: 3,
        partitionsCount: 15,
        totalMessages: 10000,
      };
      mockFetch.mockResolvedValue(mockResponse(200, metrics));

      const client = new StreamHouseClient({
        baseUrl: 'http://localhost:8080',
        fetch: mockFetch,
      });
      const result = await client.getMetrics();

      expect(result.topicsCount).toBe(5);
      expect(result.totalMessages).toBe(10000);
    });
  });
});
